src/abstract_domains.rs

```diff
@@ -3,8 +3,38 @@
 // This source code is licensed under the MIT license found in the
 // LICENSE file in the root directory of this source tree.
 
+use constant_value::ConstantValue;
+
+// See https://github.com/facebookexperimental/MIRAI/blob/master/documentation/AbstractValues.md.
+
+/// A collection of abstract domains associated with a particular abstract value.
+/// The expression domain captures the precise expression that resulted in the abstract
+/// value. It can be used to derive any other kind of abstract domain on demand.
 #[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Hash)]
 pub struct AbstractDomains {
-    //todo: just store the expression (as a domain)
-//todo: use cached getters to get other domains on demand
+    pub expression_domain: ExpressionDomain,
+    //todo: use cached getters to get other domains on demand
+}
+
+/// The things all abstract domain instances have to be able to do
+pub trait AbstractDomain {
+    //todo: join, is_top, is_bottom, etc.
+}
+
+/// An abstract domain uses a functional (side effect free) expression that precisely tracks
+/// a single value.
+#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Hash)]
+pub enum ExpressionDomain {
+    /// An expression that represents any possible value
+    Top,
+
+    /// An expression that represents an impossible value, such as the value returned by function
+    /// that always panics.
+    Bottom,
```

Comment:
Probably overkill at this point, but: in Infer we had functors for lifting a value`CompileTimeConstant(ConstantValue)` to an `AbstractDomain` by adding artificial top and bottom elements that are automatically incorporated into the partial order operations. 

I haven't figured out how to do this well in Rust yet--the lack of higher-kinded types means there's no direct analogue of the functor approach. On the other hand, Rust macros might be powerful enough to do this, but tend to make code very hard to read...

Reply:
I am very much opposed to macros and not very fond of higher-kinded types. For me, code readability is the only thing that matters and I find such code hard to read.